src/main/java/de/tum/in/www1/artemis/service/connectors/gitlabci/GitLabCIService.java

```diff
@@ -0,0 +1,222 @@
+import de.tum.in.www1.artemis.service.dto.AbstractBuildResultNotificationDTO;
+
+@Profile("gitlabci")
+@Service
+public class GitLabCIService extends AbstractContinuousIntegrationService {
+
+    private static final Logger log = LoggerFactory.getLogger(GitLabCIService.class);
+
+    // TODO: Generate an URL and make it accessible via the API
+    public static final String CI_CONFIG_URL = "https://home.in.tum.de/~scbe/Artemis/gitlabci-runners/getting-started-template.yml";
+
+    private final GitLabApi gitlab;
+
+    private final UrlService urlService;
+
+    @Value("${artemis.version-control.url}")
+    private URL gitlabServerUrl;
+
+    public GitLabCIService(ProgrammingSubmissionRepository programmingSubmissionRepository, FeedbackRepository feedbackRepository, BuildLogEntryService buildLogService,
+            RestTemplate restTemplate, RestTemplate shortTimeoutRestTemplate, GitLabApi gitlab, UrlService urlService) {
+        super(programmingSubmissionRepository, feedbackRepository, buildLogService, restTemplate, shortTimeoutRestTemplate);
+        this.gitlab = gitlab;
+        this.urlService = urlService;
+    }
+
+    @Override
+    public void createBuildPlanForExercise(ProgrammingExercise exercise, String planKey, VcsRepositoryUrl repositoryURL, VcsRepositoryUrl testRepositoryURL,
+            VcsRepositoryUrl solutionRepositoryURL) {
+        addBuildPlanToGitLabRepositoryConfiguration(repositoryURL);
+    }
+
+    private void addBuildPlanToGitLabRepositoryConfiguration(VcsRepositoryUrl repositoryURL) {
+        final String repositoryPath = getRepositoryPath(repositoryURL);
+        try {
+            Project project = gitlab.getProjectApi().getProject(repositoryPath);
+
+            project.setJobsEnabled(true);
+            project.setSharedRunnersEnabled(true);
+            project.setAutoDevopsEnabled(false);
+            project.setCiConfigPath(GitLabCIService.CI_CONFIG_URL);
+
+            gitlab.getProjectApi().updateProject(project);
+        }
+        catch (GitLabApiException e) {
+            throw new GitLabCIException("Error getting project from repository path for " + repositoryURL.toString());
+        }
+    }
+
+    private String getRepositoryPath(VcsRepositoryUrl repositoryUrl) {
+        return urlService.getRepositoryPathFromRepositoryUrl(repositoryUrl);
+    }
+
+    @Override
+    public void recreateBuildPlansForExercise(ProgrammingExercise exercise) {
+        addBuildPlanToGitLabRepositoryConfiguration(exercise.getRepositoryURL(RepositoryType.TEMPLATE));
+        addBuildPlanToGitLabRepositoryConfiguration(exercise.getRepositoryURL(RepositoryType.SOLUTION));
+    }
+
+    @Override
+    public String copyBuildPlan(String sourceProjectKey, String sourcePlanName, String targetProjectKey, String targetProjectName, String targetPlanName,
+            boolean targetProjectExists) {
+        log.error("Unsupported action: GitLabCIService.copyBuildPlan()");
```

Comment:
Is logging this enough? Consider throwing an exception.

Reply:
I wasn't sure how to handle an unsupported action, but throwing an exception sounds better because - if I did everything right - then the user will see an error in the client and not just in the logs. I will change it as soon as I have time.

Reply:
@Strohgelaender I checked whether it works with an exception, but it fails. I think when participating in an exercise, `copyBuildPlan()` is executed automatically. We just want to notify, that this is not supported. Even though, overall it works since additionally `addBuildPlanToGitLabRepositoryConfiguration()` is executed.